src/query/service/src/pipelines/builders/copy.rs

```diff
@@ -117,74 +180,66 @@ pub fn build_append_data_with_finish_pipeline(
         )?;
     }
 
-    let stage_info_clone = plan_stage_table_info.stage_info;
+    let stage_info = plan_stage_table_info.stage_info;
+    // Only append to table, not commit.
     let write_mode = plan_write_mode;
-    let mut purge = true;
+    let mut purge = false;
     match write_mode {
-        CopyIntoTableMode::Insert { overwrite } => {
-            if use_commit {
-                build_append2table_pipeline(
-                    ctx.clone(),
-                    main_pipeline,
-                    to_table,
-                    plan_required_values_schema,
-                    None,
-                    overwrite,
-                    AppendMode::Copy,
-                )?;
-            } else {
-                build_append2table_without_commit_pipeline(
-                    ctx.clone(),
-                    main_pipeline,
-                    to_table,
-                    plan_required_values_schema,
-                    AppendMode::Copy,
-                )?
-            }
-        }
+        CopyIntoTableMode::Insert { overwrite: _ } => build_append2table_without_commit_pipeline(
+            ctx.clone(),
+            main_pipeline,
+            to_table,
+            plan_required_values_schema,
+            AppendMode::Copy,
+        )?,
         CopyIntoTableMode::Copy => {
-            if !stage_info_clone.copy_options.purge {
-                purge = false;
+            if stage_info.copy_options.purge {
+                purge = true;
             }
 
-            if use_commit {
-                let copied_files = build_upsert_copied_files_to_meta_req(
-                    ctx.clone(),
-                    to_table.clone(),
-                    stage_info_clone.clone(),
-                    files.clone(),
-                    plan_force,
-                )?;
-                build_append2table_pipeline(
-                    ctx.clone(),
-                    main_pipeline,
-                    to_table,
-                    plan_required_values_schema,
-                    copied_files,
-                    false,
-                    AppendMode::Copy,
-                )?;
-            } else {
-                build_append2table_without_commit_pipeline(
-                    ctx.clone(),
-                    main_pipeline,
-                    to_table,
-                    plan_required_values_schema,
-                    AppendMode::Copy,
-                )?
-            }
+            build_append2table_without_commit_pipeline(
+                ctx.clone(),
+                main_pipeline,
+                to_table,
+                plan_required_values_schema,
+                AppendMode::Copy,
+            )?
         }
         CopyIntoTableMode::Replace => {}
     }
 
-    if local_id == ctx.get_cluster().local_id {
+    // set finished callback to pipeline.
+    set_pipeline_finish_callback(
```

Comment:
please make sure the pruge will only happen on the source node. thanks!! It means only source need to do purge operations, remote node doesn't need, thanks. other is fine for me.

Reply:
The check is move to `set_pipeline_finish_callback` , we check here:

https://github.com/datafuselabs/databend/pull/12031/files#diff-8f4130db0584112576635553cf35d4760ea25d2e3740ba2de267c77ab913ce2eR233